src/project.ts

```
@@ -0,0 +1,110 @@
+      const p = path.join(juliaProject, filename)
+      if (fs.existsSync(p) && fs.lstatSync(p).isFile()) {
+        juliaProjectFile = p
+        break
+      }
+    }
+  }
+
+  if (!juliaProjectFile) {
+    throw new Error(
+      `Unable to locate project file with project input: ${juliaProject}`
+    )
+  }
+
+  return juliaProjectFile
+}
+
+/**
+ * Determine the NPM semver range string from a parsed Julia project TOML.
+ *
+ * @returns A NPM semver range string.
+ * @throws Error if the Julia compat range cannot be converted into a NPM semver range.
+ */
+export function getJuliaCompatRange(juliaProject: JuliaProjectTOML): string {
+  let compatRange: string | null
+
+  if (juliaProject.compat?.julia !== undefined) {
+    compatRange = validJuliaCompatRange(juliaProject.compat.julia)
+  } else {
+    compatRange = "*"
+  }
+
+  if (!compatRange) {
+    throw new Error(
+      `Invalid version range found in Julia compat: ${compatRange}`
+    )
+  }
+
+  return compatRange
+}
+
+/**
+ * Convert a Julia compat range into a NPM semver range.
+ *
+ * @returns An NPM semver range string or null if the input is invalid.
+ */
+export function validJuliaCompatRange(compatRange: string): string | null {
+  const ranges: Array<string> = []
+  for (let range of compatRange.split(",")) {
+    range = range.trim()
+
+    // An empty range isn't supported by Julia
+    if (!range) {
+      return null
+    }
+
+    // NPM's semver doesn't understand unicode characters such as `≥` so we'll
+    // convert to alternatives.
+    range = range.replace("≥", ">=").replace("≤", "<=")
+
+    // Cleanup whitespace. Julia only allows whitespace between the specifier
+    // and version with certain specifiers.
+    range = range.replace(/\s+/g, " ").replace(/(?<=(>|>=|≥|<)) (?=\d)/g, "")
+
+    if (
+      !semver.validRange(range) ||
+      range.split(/(?<! -) (?!- )/).length > 1 ||
+      range.startsWith("<=") ||
+      range === "*"
+    ) {
+      return null
```

Instead of returning `null` here, would it not be better to fail earlier? Then we can simplify the `resolveVersion()` and `resolveVersions()`.

---

I'll consider that. I do think though that missing a `julia` compat entry doesn't always have to result in a failure.

---

Oh that's a good call! I was just looking at the diff here and noticed it seems to just get called in `resolveVersion()` where if `!juliaCompatRange` it would throw an error.